refactor(linter): move ignore logic to `LintIgnoreMatcher`

Sysix opened the pull request:
Moved the logic into an own struct.
`oxlint` does it only need at the beginning for prefiltering the files.
`oxc_language_server` needs to keep the ignores alive.

Before the `LintConfig` would hold it, which is too long for `oxlint`.

Boshen commented:
### Merge activity

* **Aug 26, 2:39 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Sysix committed:
refactor(linter): move ignore logic to `LintIgnoreMatcher` (#13222)

Moved the logic into an own struct.
`oxlint` does it only need at the beginning for prefiltering the files.
`oxc_language_server` needs to keep the ignores alive.

Before the `LintConfig` would hold it, which is too long for `oxlint`.
